File: client/src/components/controls/controlTypes.js
// Copyright (c) 2026 TRV Enterprises LLC
// Licensed under Apache 2.0
// See LICENSE file for details.

/**
 * Control type constants and metadata.
 * Separated from index.js to avoid circular dependencies
 * (ControlRenderer imports these, and index.js exports ControlRenderer).
 */

import {
  mdiGestureTap,
  mdiToggleSwitch,
  mdiTuneVertical,
  mdiFormTextbox,
  mdiPowerPlug,
  mdiLightbulbOn,
  mdiFormatText,
  mdiGarage,
  mdiSend
} from '@mdi/js';

// Control type constants - must match backend models.ControlType* constants
export const CONTROL_TYPES = {
  BUTTON: 'button',
  TOGGLE: 'toggle',
  SLIDER: 'slider',
  TEXT_INPUT: 'text_input',
  SWITCH: 'switch',
  DIMMER: 'dimmer',
  GARAGE_DOOR: 'garage_door',
  TILE_SWITCH: 'tile_switch',
  TILE_DIMMER: 'tile_dimmer',
  TILE_GARAGE_DOOR: 'tile_garage_door',
  MQTT_PUBLISH: 'mqtt_publish',
  TEXT_LABEL: 'text_label'
};

// Control type metadata for UI, discovery, and AI builder
export const CONTROL_TYPE_INFO = {
  [CONTROL_TYPES.BUTTON]: {
    label: 'Button',
    description: 'Simple action button that triggers a command when clicked',
    icon: mdiGestureTap,
    category: 'carbon',
    canWrite: true,
    canRead: false,
    defaultUIConfig: {
      label: 'Execute',
      kind: 'primary'
    }
  },
  [CONTROL_TYPES.TOGGLE]: {
    label: 'Toggle',
    description: 'On/off switch that sends true or false',
    icon: mdiToggleSwitch,
    category: 'carbon',
    canWrite: true,
    canRead: true,
    defaultUIConfig: {
      label: 'Enable',
      offLabel: 'Disable',
      state_field: 'state'
    }
  },
  [CONTROL_TYPES.SLIDER]: {
    label: 'Slider',
    description: 'Numeric slider for setting values within a range',
    icon: mdiTuneVertical,
    category: 'carbon',
    canWrite: true,
    canRead: true,
    defaultUIConfig: {
      label: 'Value',
      min: 0,
      max: 100,
      step: 1,
      state_field: 'value'
    }
  },
  [CONTROL_TYPES.TEXT_INPUT]: {
    label: 'Text Input',
    description: 'Text field for entering custom values or commands',
    icon: mdiFormTextbox,
    category: 'carbon',
    canWrite: true,
    canRead: true,
    defaultUIConfig: {
      label: 'Command',
      placeholder: 'Enter value...',
      submitLabel: 'Send',
      state_field: 'value'
    }
  },
  [CONTROL_TYPES.SWITCH]: {
    label: 'Switch',
    description: 'On/off switch with HomeKit-style pill design',
    icon: mdiPowerPlug,
    category: 'custom',
    canWrite: true,
    canRead: true,
    defaultUIConfig: {
      label: 'Switch',
      onLabel: 'On',
      offLabel: 'Off',
      state_field: 'state'
    }
  },
  [CONTROL_TYPES.DIMMER]: {
    label: 'Dimmer',
    description: 'Vertical slider for dimming lights, drag to set level',
    icon: mdiLightbulbOn,
    category: 'custom',
    canWrite: true,
    canRead: true,
    defaultUIConfig: {
      label: 'Light',
      min: 0,
      max: 100,
      step: 1,
      state_field: 'level'
    }
  },
  [CONTROL_TYPES.GARAGE_DOOR]: {
    label: 'Garage Door',
    description: 'Full-size animated garage door with open/closed state from a contact sensor.',
    icon: mdiGarage,
    category: 'custom',
    canWrite: false,
    canRead: true,
    defaultUIConfig: {
      label: 'Garage',
      state_field: 'contact'
    }
  },
  [CONTROL_TYPES.TILE_SWITCH]: {
    label: 'Tile Switch',
    description: 'Compact tile showing on/off state. Tap to open full control.',
    icon: mdiPowerPlug,
    category: 'tile',
    canWrite: true,
    canRead: true,
    defaultUIConfig: {
      label: 'Switch',
      onLabel: 'On',
      offLabel: 'Off',
      state_field: 'state'
    }
  },
  [CONTROL_TYPES.TILE_DIMMER]: {
    label: 'Tile Dimmer',
    description: 'Compact tile showing dimmer level with vertical fill. Tap to open full control.',
    icon: mdiLightbulbOn,
    category: 'tile',
    canWrite: true,
    canRead: true,
    defaultUIConfig: {
      label: 'Light',
      min: 0,
      max: 100,
      step: 1,
      state_field: 'level'
    }
  },
  [CONTROL_TYPES.TILE_GARAGE_DOOR]: {
    label: 'Tile Garage Door',
    description: 'Compact tile showing garage door open/closed state from a contact sensor.',
    icon: mdiGarage,
    category: 'tile',
    canWrite: false,
    canRead: true,
    defaultUIConfig: {
      label: 'Garage',
      state_field: 'contact'
    }
  },
  [CONTROL_TYPES.MQTT_PUBLISH]: {
    label: 'MQTT Publish',
    description: 'Fire-and-forget button that publishes a static JSON payload to an MQTT topic when pressed. No device type needed.',
    icon: mdiSend,
    category: 'carbon',
    canWrite: true,
    canRead: false,
    defaultUIConfig: {
      label: 'Publish',
      kind: 'primary',
      payload: {}
    }
  },
  [CONTROL_TYPES.TEXT_LABEL]: {
    label: 'Text Label',
    description: 'Static text display for section headers, date/time, or dashboard titles. No connection needed.',
    icon: mdiFormatText,
    category: 'decorative',
    hidden: true, // Replaced by native text panels — kept for backward compat
    canWrite: false,
    canRead: false,
    defaultUIConfig: {
      display_content: 'title',
      align: 'center',
      size: 'md'
    }
  },

  // --- Backward-compat aliases ---
  // Legacy DB records may still have these control types. The renderers
  // self-register both old and new names (see ControlPlug.jsx, TilePlug.jsx),
  // but the editor also needs metadata entries so the Connection dropdown
  // and other editor fields render correctly. Hidden from the type selector
  // so users can't create new ones — they only appear when editing legacy
  // components.
  plug: {
    label: 'Plug (legacy)',
    description: 'Legacy name for Switch. Kept so existing components can still be edited.',
    icon: mdiPowerPlug,
    category: 'custom',
    hidden: true,
    canWrite: true,
    canRead: true,
    defaultUIConfig: {
      label: 'Switch',
      onLabel: 'On',
      offLabel: 'Off',
      state_field: 'state'
    }
  },
  tile_plug: {
    label: 'Tile Plug (legacy)',
    description: 'Legacy name for Tile Switch. Kept so existing components can still be edited.',
    icon: mdiPowerPlug,
    category: 'tile',
    hidden: true,
    canWrite: true,
    canRead: true,
    defaultUIConfig: {
      label: 'Switch',
      onLabel: 'On',
      offLabel: 'Off',
      state_field: 'state'
    }
  }
};

// Categories for widget selector grouping
export const CONTROL_CATEGORIES = {
  carbon: { label: 'Carbon Controls', description: 'Standard Carbon Design System components' },
  custom: { label: 'Custom Controls', description: 'Full-size custom controls for switches and dimmers' },
  tile: { label: 'Tiles', description: 'Compact state tiles for dashboards' }
};

File: server-go/internal/registry/control_types.go
// Copyright (c) 2026 TRV Enterprises LLC
// Licensed under Apache 2.0
// See LICENSE file for details.

package registry

// Control type registrations. This is the backend mirror of the frontend
// controlTypes.js CONTROL_TYPE_INFO map. When adding or editing a control
// type you must update both files — the frontend for rendering, this file
// for discovery (AI builder, MCP server, /api/meta/types).
//
// Eventually this will flip: control type metadata will come from the
// registry (optionally MongoDB-backed) and the frontend will fetch it at
// runtime. The frontend React component itself will always be code.
//
// Icons are stored as MDI constant names ("mdiGestureTap", not "gesture-tap").
// The frontend resolves them to SVG paths at render time via @mdi/js.

func init() {
	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.button",
		Category:    CategoryControl,
		Subtype:     "button",
		DisplayName: "Button",
		Description: "Simple action button that triggers a command when clicked.",
		Icon:        "mdiGestureTap",
		UICategory:  "carbon",
		Capabilities: ComponentCapabilities{
			CanWrite:           true,
			RequiresDeviceType: true,
			RequiresConnection: true,
		},
		DefaultConfig: map[string]interface{}{
			"label": "Execute",
			"kind":  "primary",
		},
	})

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.toggle",
		Category:    CategoryControl,
		Subtype:     "toggle",
		DisplayName: "Toggle",
		Description: "On/off switch that sends true or false and subscribes to live state.",
		Icon:        "mdiToggleSwitch",
		UICategory:  "carbon",
		Capabilities: ComponentCapabilities{
			CanRead:            true,
			CanWrite:           true,
			RequiresDeviceType: true,
			RequiresConnection: true,
			SupportsStreaming:  true,
		},
		DefaultConfig: map[string]interface{}{
			"label":       "Enable",
			"offLabel":    "Disable",
			"state_field": "state",
		},
	})

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.slider",
		Category:    CategoryControl,
		Subtype:     "slider",
		DisplayName: "Slider",
		Description: "Numeric slider for setting values within a range. Reads live state and writes command values.",
		Icon:        "mdiTuneVertical",
		UICategory:  "carbon",
		Capabilities: ComponentCapabilities{
			CanRead:            true,
			CanWrite:           true,
			RequiresDeviceType: true,
			RequiresConnection: true,
			SupportsStreaming:  true,
		},
		DefaultConfig: map[string]interface{}{
			"label":       "Value",
			"min":         0,
			"max":         100,
			"step":        1,
			"state_field": "value",
		},
	})

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.text_input",
		Category:    CategoryControl,
		Subtype:     "text_input",
		DisplayName: "Text Input",
		Description: "Text field for entering custom values or commands.",
		Icon:        "mdiFormTextbox",
		UICategory:  "carbon",
		Capabilities: ComponentCapabilities{
			CanRead:            true,
			CanWrite:           true,
			RequiresDeviceType: true,
			RequiresConnection: true,
			SupportsStreaming:  true,
		},
		DefaultConfig: map[string]interface{}{
			"label":       "Command",
			"placeholder": "Enter value...",
			"submitLabel": "Send",
			"state_field": "value",
		},
	})

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.switch",
		Category:    CategoryControl,
		Subtype:     "switch",
		DisplayName: "Switch",
		Description: "On/off switch with HomeKit-style pill design.",
		Icon:        "mdiPowerPlug",
		UICategory:  "custom",
		Capabilities: ComponentCapabilities{
			CanRead:            true,
			CanWrite:           true,
			RequiresDeviceType: true,
			RequiresConnection: true,
			SupportsStreaming:  true,
		},
		DefaultConfig: map[string]interface{}{
			"label":       "Switch",
			"onLabel":     "On",
			"offLabel":    "Off",
			"state_field": "state",
		},
	})

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.dimmer",
		Category:    CategoryControl,
		Subtype:     "dimmer",
		DisplayName: "Dimmer",
		Description: "Vertical slider for dimming lights — drag to set level.",
		Icon:        "mdiLightbulbOn",
		UICategory:  "custom",
		Capabilities: ComponentCapabilities{
			CanRead:            true,
			CanWrite:           true,
			RequiresDeviceType: true,
			RequiresConnection: true,
			SupportsStreaming:  true,
		},
		DefaultConfig: map[string]interface{}{
			"label":       "Light",
			"min":         0,
			"max":         100,
			"step":        1,
			"state_field": "level",
		},
	})

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.garage_door",
		Category:    CategoryControl,
		Subtype:     "garage_door",
		DisplayName: "Garage Door",
		Description: "Full-size animated garage door showing open/closed state from a contact sensor. Read-only.",
		Icon:        "mdiGarage",
		UICategory:  "custom",
		Capabilities: ComponentCapabilities{
			CanRead:            true,
			RequiresConnection: true,
			SupportsStreaming:  true,
		},
		DefaultConfig: map[string]interface{}{
			"label":       "Garage",
			"state_field": "contact",
		},
	})

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.tile_switch",
		Category:    CategoryControl,
		Subtype:     "tile_switch",
		DisplayName: "Tile Switch",
		Description: "Compact tile showing on/off state. Tap to open the full control.",
		Icon:        "mdiPowerPlug",
		UICategory:  "tile",
		Capabilities: ComponentCapabilities{
			CanRead:            true,
			CanWrite:           true,
			RequiresDeviceType: true,
			RequiresConnection: true,
			SupportsStreaming:  true,
		},
		DefaultConfig: map[string]interface{}{
			"label":       "Switch",
			"onLabel":     "On",
			"offLabel":    "Off",
			"state_field": "state",
		},
	})

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.tile_dimmer",
		Category:    CategoryControl,
		Subtype:     "tile_dimmer",
		DisplayName: "Tile Dimmer",
		Description: "Compact tile showing dimmer level with a vertical fill. Tap to open the full control.",
		Icon:        "mdiLightbulbOn",
		UICategory:  "tile",
		Capabilities: ComponentCapabilities{
			CanRead:            true,
			CanWrite:           true,
			RequiresDeviceType: true,
			RequiresConnection: true,
			SupportsStreaming:  true,
		},
		DefaultConfig: map[string]interface{}{
			"label":       "Light",
			"min":         0,
			"max":         100,
			"step":        1,
			"state_field": "level",
		},
	})

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.tile_garage_door",
		Category:    CategoryControl,
		Subtype:     "tile_garage_door",
		DisplayName: "Tile Garage Door",
		Description: "Compact tile showing garage door open/closed state from a contact sensor. Read-only.",
		Icon:        "mdiGarage",
		UICategory:  "tile",
		Capabilities: ComponentCapabilities{
			CanRead:            true,
			RequiresConnection: true,
			SupportsStreaming:  true,
		},
		DefaultConfig: map[string]interface{}{
			"label":       "Garage",
			"state_field": "contact",
		},
	})

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.mqtt_publish",
		Category:    CategoryControl,
		Subtype:     "mqtt_publish",
		DisplayName: "MQTT Publish",
		Description: "Fire-and-forget button that publishes a static JSON payload to an MQTT topic when pressed. No device type needed — topic and payload are configured directly on the control.",
		Icon:        "mdiSend",
		UICategory:  "carbon",
		Capabilities: ComponentCapabilities{
			CanWrite:           true,
			RequiresConnection: true,
		},
		DefaultConfig: map[string]interface{}{
			"label":   "Publish",
			"kind":    "primary",
			"payload": map[string]interface{}{},
		},
	})

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.text_label",
		Category:    CategoryControl,
		Subtype:     "text_label",
		DisplayName: "Text Label",
		Description: "Static text display for section headers, date/time, or dashboard titles. No connection needed. Replaced by native text panels — kept for backward compat.",
		Icon:        "mdiFormatText",
		UICategory:  "decorative",
		Hidden:      true,
		Capabilities: ComponentCapabilities{},
		DefaultConfig: map[string]interface{}{
			"display_content": "title",
			"align":           "center",
			"size":            "md",
		},
	})

	// --- Backward-compat aliases ---
	// Legacy records may still have these control types. Kept hidden so new
	// components can't be created with them, but the editor and MCP still
	// need metadata entries to load existing records cleanly.

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.plug",
		Category:    CategoryControl,
		Subtype:     "plug",
		DisplayName: "Plug (legacy)",
		Description: "Legacy name for Switch. Kept so existing components can still be edited.",
		Icon:        "mdiPowerPlug",
		UICategory:  "custom",
		Hidden:      true,
		Capabilities: ComponentCapabilities{
			CanRead:            true,
			CanWrite:           true,
			RequiresDeviceType: true,
			RequiresConnection: true,
			SupportsStreaming:  true,
		},
		DefaultConfig: map[string]interface{}{
			"label":       "Switch",
			"onLabel":     "On",
			"offLabel":    "Off",
			"state_field": "state",
		},
	})

	RegisterComponentType(ComponentTypeInfo{
		TypeID:      "control.tile_plug",
		Category:    CategoryControl,
		Subtype:     "tile_plug",
		DisplayName: "Tile Plug (legacy)",
		Description: "Legacy name for Tile Switch. Kept so existing components can still be edited.",
		Icon:        "mdiPowerPlug",
		UICategory:  "tile",
		Hidden:      true,
		Capabilities: ComponentCapabilities{
			CanRead:            true,
			CanWrite:           true,
			RequiresDeviceType: true,
			RequiresConnection: true,
			SupportsStreaming:  true,
		},
		DefaultConfig: map[string]interface{}{
			"label":       "Switch",
			"onLabel":     "On",
			"offLabel":    "Off",
			"state_field": "state",
		},
	})
}
